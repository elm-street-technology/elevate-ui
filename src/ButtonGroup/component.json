{
  "description": "A component used to render a group of buttons that can have a single or multiple values.",
  "displayName": "ButtonGroup",
  "methods": [
    {
      "name": "onChange",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "e",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "handleClick",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "input",
          "type": null
        }
      ],
      "returns": null
    }
  ],
  "props": {
    "classes": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "className": {
      "flowType": {
        "name": "string"
      },
      "required": true,
      "description": "classNames to be passed to the component."
    },
    "field": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "form": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "items": {
      "flowType": {
        "name": "Array",
        "elements": [
          {
            "name": "signature",
            "type": "object",
            "raw": "{ label: string, value: string, disabled: boolean }",
            "signature": {
              "properties": [
                {
                  "key": "label",
                  "value": {
                    "name": "string",
                    "required": true
                  }
                },
                {
                  "key": "value",
                  "value": {
                    "name": "string",
                    "required": true
                  }
                },
                {
                  "key": "disabled",
                  "value": {
                    "name": "boolean",
                    "required": true
                  }
                }
              ]
            }
          }
        ],
        "raw": "Array<{ label: string, value: string, disabled: boolean }>"
      },
      "required": false,
      "description": "Array of items to use as the values. `{label: string, value: string}`",
      "defaultValue": {
        "value": "[]",
        "computed": false
      }
    },
    "label": {
      "flowType": {
        "name": "string"
      },
      "required": true,
      "description": "Text input to be used as the label for the checkbox group."
    },
    "theme": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "multiSelect": {
      "flowType": {
        "name": "boolean"
      },
      "required": false,
      "description": "Allows the button group to have multiple values selected when set to true",
      "defaultValue": {
        "value": "false",
        "computed": false
      }
    },
    "color": {
      "flowType": {
        "name": "string"
      },
      "required": false,
      "description": "Color to be applied to the Button component. Can be \"primary\", \"secondary\", or any custom value.",
      "defaultValue": {
        "value": "\"primary\"",
        "computed": false
      }
    }
  }
}