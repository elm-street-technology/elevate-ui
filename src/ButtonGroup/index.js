import React, { Component, Fragment } from "react";
import withStyles from "react-jss";
import classNames from "classnames";
import Label from "../Label";
import without from "lodash/without";
import Color from "color";

type Props = {
  classes: Object,
  /**
   * classNames to be passed to the component.
   */
  className: string,
  field: Object, // needs flow-typed https://github.com/flowtype/flow-typed/issues/1903
  form: Object, // needs flow-typed https://github.com/flowtype/flow-typed/issues/1903
  /**
   * Array of items to use as the values. `{label: string, value: string}`
   */
  items: Array<{ label: string, value: string, disabled: boolean }>,
  /**
   * Text input to be used as the label for the checkbox group.
   */
  label: string,
  theme: Object,
  /**
   * Allows the button group to have multiple values selected when set to true
   */
  multiSelect?: boolean,
  /**
   * Color to be applied to the Button component. Can be "primary", "secondary", or any custom value.
   */
  color?: string,
};

function getBackgroundColor(theme, props) {
  if (props.color !== "primary" && props.color !== "secondary") {
    return props.color;
  } else {
    return theme.colors[props.color];
  }
}

function getSpanColor(theme, props) {
  let backgroundColor;
  if (props.color === "primary" || props.color === "secondary") {
    backgroundColor = theme.colors[props.color];
  } else {
    backgroundColor = props.color;
  }

  if (Color(backgroundColor).isDark()) {
    return theme.colors["white"];
  } else if (Color(backgroundColor).isLight()) {
    return theme.colors["gray900"];
  } else {
    return theme.colors["gray500"];
  }
}

/**
 * A component used to render a group of buttons that can have a single or multiple values.
 */
class ButtonGroup extends Component<Props> {
  static defaultProps = {
    items: [],
    multiSelect: false,
    color: "primary",
  };

  onChange = (e) => {
    const {
      field: { name, value },
      form: { setFieldValue },
    } = this.props;
    const selectedButton = e.target.id;
    if (this.props.multiSelect) {
      let selectedButtons = Array.isArray(value) ? value.slice() : [];
      if (e.target.checked && selectedButtons.indexOf(selectedButton) === -1) {
        selectedButtons.push(selectedButton);
      } else if (!e.target.checked) {
        selectedButtons = without(selectedButtons, selectedButton);
      }
      return setFieldValue(name, selectedButtons);
    } else {
      if (value === selectedButton) {
        return setFieldValue(name, "");
      }

      return setFieldValue(name, selectedButton);
    }
  };

  render() {
    const {
      label,
      items,
      classes,
      field: { name, value },
      multiSelect,
      form: { values },
    } = this.props;

    return (
      <div className={classes.scaffold}>
        {label && <Label>{label}</Label>}
        <div className={classNames(classes.toggles, classes.inline)}>
          {items.map((item) => {
            const input = item.value;
            return (
              <Fragment key={item.value}>
                <label
                  className={classNames(
                    classes.button,
                    Array.isArray(values[name])
                      ? values[name].includes(item.value)
                        ? classes.buttonActive
                        : null
                      : values[name] === item.value
                        ? classes.buttonActive
                        : null
                  )}
                >
                  <span
                    className={
                      this[`input${item.value}`] &&
                      this[`input${item.value}`].checked
                        ? classes.spanTextActive
                        : classes.spanText
                    }
                  >
                    {item.label}
                  </span>
                  <input
                    id={item.value}
                    type="checkbox"
                    name={name}
                    className={classes.hiddenInput}
                    checked={
                      multiSelect
                        ? value.indexOf(item.value) > -1
                        : value === item.value
                    }
                    value={item.value}
                    onChange={this.onChange}
                    ref={(node) => (this[`input${item.value}`] = node)}
                  />
                </label>
              </Fragment>
            );
          })}
        </div>
      </div>
    );
  }
}

export default withStyles((theme) => ({
  scaffold: {
    margin: "8px auto 16px",
  },
  toggles: {
    display: "flex",
    flexDirection: "column",
    flexWrap: "wrap",
    alignItems: "flex-start",
    marginLeft: "-12px",
  },
  inline: {
    flexDirection: "row",
  },
  hiddenInput: {
    display: "none",
  },
  button: {
    background: "none",
    border: `1px solid ${theme.colors["gray200"]}`,
    borderRight: "none",
    padding: 12,
    cursor: "pointer",
    "&:first-of-type": {
      borderLeft: `1px solid ${theme.colors["gray200"]}`,
      borderTopLeftRadius: 5,
      borderBottomLeftRadius: 5,
    },
    "&:last-of-type": {
      borderRight: `1px solid ${theme.colors["gray200"]}`,
      borderTopRightRadius: 5,
      borderBottomRightRadius: 5,
    },
  },
  spanText: {
    pointerEvents: "none",
    color: theme.colors["gray500"],
  },
  spanTextActive: {
    pointerEvents: "none",
    color: (props) => getSpanColor(theme, props),
  },
  buttonActive: {
    background: (props) => getBackgroundColor(theme, props),
    border: `1px solid ${theme.colors["gray200"]}`,
    borderRight: "none",
    padding: 12,
    cursor: "pointer",
    "&:first-of-type": {
      borderLeft: `1px solid ${theme.colors["gray200"]}`,
      borderTopLeftRadius: 5,
      borderBottomLeftRadius: 5,
    },
    "&:last-of-type": {
      borderRight: `1px solid ${theme.colors["gray200"]}`,
      borderTopRightRadius: 5,
      borderBottomRightRadius: 5,
    },
  },
}))(ButtonGroup);
