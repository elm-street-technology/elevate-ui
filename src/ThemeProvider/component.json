{
  "description": "",
  "displayName": "ThemeProvider",
  "methods": [],
  "props": {
    "withReset": {
      "defaultValue": {
        "value": "true",
        "computed": false
      },
      "required": false
    }
  }
}