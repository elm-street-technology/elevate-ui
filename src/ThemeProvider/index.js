// @flow
import React from "react";
import injectSheet, { ThemeProvider as JSSThemeProvider } from "react-jss";
import merge from "lodash/merge";

const colors = {
  primary: "#f15953",
  primaryDark: "#d24f45",
  primaryGlow: "#d24f45", // todo
  secondary: "#55c3ba",
  secondaryDark: "#44b0a7",
  tertiary: "#2e7fc2",
  quaternary: "#ffc629",
  white: "#fff",
  gray050: "#fbfcfd",
  gray100: "#edf1f4",
  gray200: "#dde2e7",
  gray300: "#ccd2d8",
  gray400: "#b3bac1",
  gray500: "#888f96",
  gray600: "#60686f",
  gray700: "#4b565f",
  gray800: "#5a5b5c",
  gray900: "#232c35",
  black: "#11181e",
  error: "#9c2929",
  facebook: "#3b5998",
  google: "#db4437",
  linkedin: "#007bb5",
  pinterest: "#bd081c",
  twitter: "#1da1f2",
};

const alertColors = {
  error: {
    color: "#9c2929",
    backgroundColor: "#fcf2f2",
    borderColor: "#f6dddd",
  },
  info: {
    color: "#0c5460",
    backgroundColor: "#d1ecf1",
    borderColor: "#bee5eb",
  },
  success: {
    color: "#246761",
    backgroundColor: "#ecf8f7",
    borderColor: "#d2e8e6",
  },
  warning: {
    color: "#856404",
    backgroundColor: "#fff3cd",
    borderColor: "#ffeeba",
  },
};

const typography = {
  bodyColor: "#2E2E35",
  fontFamily:
    "'Open Sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "1rem",
};

const transitions = {
  default: "all 200ms linear",
};

const breakpoints = (minWidth: number) => {
  if (!minWidth) {
    return new Error("Must enter a valid minWidth to use theme breakpoints");
  }
  return `@media (min-width: ${minWidth}px)`;
};

const globalBorderRadius = "4px";
const globalBoxShadow =
  "0 2px 5px 0 rgba(0,0,0,0.16), 0 2px 10px 0 rgba(0,0,0,0.12)";
const globalPadding = {
  padding: "12px",
  [breakpoints.tabletPortrait]: {
    padding: "16px",
  },
  [breakpoints.tabletLandscape]: {
    padding: "20px",
  },
  [breakpoints.desktop]: {
    padding: "24px",
  },
};

const shadowKeyUmbraOpacity = 0.2;
const shadowKeyPenumbraOpacity = 0.14;
const shadowAmbientShadowOpacity = 0.12;

function createShadow(...px) {
  return [
    `${px[0]}px ${px[1]}px ${px[2]}px ${
      px[3]
    }px rgba(0, 0, 0, ${shadowKeyUmbraOpacity})`,
    `${px[4]}px ${px[5]}px ${px[6]}px ${
      px[7]
    }px rgba(0, 0, 0, ${shadowKeyPenumbraOpacity})`,
    `${px[8]}px ${px[9]}px ${px[10]}px ${
      px[11]
    }px rgba(0, 0, 0, ${shadowAmbientShadowOpacity})`,
  ].join(",");
}

const shadows = [
  createShadow(0, 1, 3, 0, 0, 1, 1, 0, 0, 2, 1, -1),
  createShadow(0, 1, 5, 0, 0, 2, 2, 0, 0, 3, 1, -2),
  createShadow(0, 1, 8, 0, 0, 3, 4, 0, 0, 3, 3, -2),
  createShadow(0, 2, 4, -1, 0, 4, 5, 0, 0, 1, 10, 0),
  createShadow(0, 3, 5, -1, 0, 5, 8, 0, 0, 1, 14, 0),
  createShadow(0, 3, 5, -1, 0, 6, 10, 0, 0, 1, 18, 0),
  createShadow(0, 4, 5, -2, 0, 7, 10, 1, 0, 2, 16, 1),
  createShadow(0, 5, 5, -3, 0, 8, 10, 1, 0, 3, 14, 2),
  createShadow(0, 5, 6, -3, 0, 9, 12, 1, 0, 3, 16, 2),
  createShadow(0, 6, 6, -3, 0, 10, 14, 1, 0, 4, 18, 3),
  createShadow(0, 6, 7, -4, 0, 11, 15, 1, 0, 4, 20, 3),
  createShadow(0, 7, 8, -4, 0, 12, 17, 2, 0, 5, 22, 4),
  createShadow(0, 7, 8, -4, 0, 13, 19, 2, 0, 5, 24, 4),
  createShadow(0, 7, 9, -4, 0, 14, 21, 2, 0, 5, 26, 4),
  createShadow(0, 8, 9, -5, 0, 15, 22, 2, 0, 6, 28, 5),
  createShadow(0, 8, 10, -5, 0, 16, 24, 2, 0, 6, 30, 5),
  createShadow(0, 8, 11, -5, 0, 17, 26, 2, 0, 6, 32, 5),
  createShadow(0, 9, 11, -5, 0, 18, 28, 2, 0, 7, 34, 6),
  createShadow(0, 9, 12, -6, 0, 19, 29, 2, 0, 7, 36, 6),
  createShadow(0, 10, 13, -6, 0, 20, 31, 3, 0, 8, 38, 7),
  createShadow(0, 10, 13, -6, 0, 21, 33, 3, 0, 8, 40, 7),
  createShadow(0, 10, 14, -6, 0, 22, 35, 3, 0, 8, 42, 7),
  createShadow(0, 11, 14, -7, 0, 23, 36, 3, 0, 9, 44, 8),
  createShadow(0, 11, 15, -7, 0, 24, 38, 3, 0, 9, 46, 8),
];

const zIndex = {
  buttonBase: 0,
  buttonChildren: 1,
  selectDropdown: 2,
  tableLoading: -1,
  tableLoadingActive: 2,
  tableResizer: 10,
  tooltip: 1000,
  modal: 2000,
};

const defaultTheme = {
  alertColors,
  breakpoints,
  colors,
  globalBorderRadius,
  globalBoxShadow,
  globalPadding,
  shadows,
  transitions,
  typography,
  zIndex,
};

const GlobalsAndReset = injectSheet((theme) => ({
  "@global": {
    /* Eric Meyer Reset v2.0 */
    /*Â https://meyerweb.com/eric/tools/css/reset/ */
    "html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video": {
      margin: "0",
      padding: "0",
      border: "0",
      fontSize: "100%",
      font: "inherit",
      verticalAlign: "baseline",
    },
    "article, aside, details, figcaption, figure, footer, header, hgroup, menu, nav, section": {
      display: "block",
    },
    body: {
      lineHeight: "1",
    },
    "ol, ul": {
      listStyle: "none",
    },
    "blockquote, q": {
      quotes: "none",
    },
    "blockquote:before, blockquote:after, q:before, q:after": {
      content: "''",
      content: "none", // eslint-disable-line
    },
    table: {
      borderCollapse: "collapse",
      borderSpacing: "0",
    },
    /* Globals */
    "html, body": {
      width: "100%",
      minHeight: "100%",
      fontFamily: theme.typography.fontFamily || "sans-serif",
      fontSize: theme.typography.fontSize || "1rem",
      fontWeight: "500",
      color: theme.typography.bodyColor || "#000",
      backgroundColor: "#fafafa",
      "-webkit-font-smoothing": "antialiased",
      "-moz-osx-font-smoothing": "grayscale",
      margin: "0",
      padding: "0",
    },
    /* Box-sizing border-box */
    /* https://www.paulirish.com/2012/box-sizing-border-box-ftw/ */
    html: {
      boxSizing: "border-box",
      background: "transparent",
    },
    "*, *:before, *:after": {
      boxSizing: "inherit",
    },
    /* Reset `button` to nothing */
    /* https://gist.github.com/chrisheninger/a860f87ef4e529b2df606768b97665a8 */
    button: {
      display: "inline-block",
      "-webkit-appearance": "none",
      "-moz-appearance": "none",
      userSelect: "none",
      textDecoration: "none",
      fontFamily: "inherit",
      fontSize: "inherit",
      color: "inherit",
      cursor: "pointer",
      border: "none",
      backgroundColor: "transparent",
      borderColor: "transparent",
      padding: "0",
      margin: "0",
      "&:disabled": {
        opacity: 0.6,
        cursor: "not-allowed",
      },
    },
  },
}))(({ children }) => children);

const ThemeProvider = ({ children, theme, withReset = true, ...rest }: any) => {
  const mergedTheme = theme ? merge(defaultTheme, theme) : defaultTheme;
  return (
    <JSSThemeProvider theme={mergedTheme} {...rest}>
      {withReset ? <GlobalsAndReset>{children}</GlobalsAndReset> : children}
    </JSSThemeProvider>
  );
};

export default ThemeProvider;
