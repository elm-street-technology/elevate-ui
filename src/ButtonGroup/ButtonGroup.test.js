import React from "react";
import renderer from "react-test-renderer";

import ThemeProvider from "../ThemeProvider";
import ButtonGroup from "./";

const WrappedGroup = (props) => (
  <ThemeProvider>
    <ButtonGroup {...props} />
  </ThemeProvider>
);

test("renders without crashing, matches the snapshot", () => {
  const component = renderer.create(
    <WrappedGroup
      id="test-buttongroup"
      label="Test ButtonGroup"
      field={{
        name: "buttongroup",
        onChange: () => {},
        onBlur: () => {},
        value: "",
      }}
      form={{
        errors: [],
        setFieldValue: () => {},
        setFieldTouched: () => {},
        touched: false,
        values: {},
      }}
      items={[
        { value: "teal", label: "Teal" },
        { value: "brown", label: "Brown" },
        { value: "black", label: "Black" },
        { value: "pink", label: "Pink" },
      ]}
    />
  );
  const componentJSON = component.toJSON();
  expect(componentJSON).toMatchSnapshot();
});
