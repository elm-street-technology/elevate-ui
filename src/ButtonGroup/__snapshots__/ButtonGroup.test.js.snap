// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing, matches the snapshot 1`] = `
<div
  className="ButtonGroup-scaffold-0-1-1"
>
  <label
    className="Label-root-0-1-11"
  >
    Test ButtonGroup
  </label>
  <div
    className="ButtonGroup-toggles-0-1-2 ButtonGroup-inline-0-1-3"
  >
    <label
      className="ButtonGroup-button-0-1-5"
    >
      <span
        className="ButtonGroup-spanText-0-1-6"
      >
        Teal
      </span>
      <input
        checked={false}
        className="ButtonGroup-hiddenInput-0-1-4"
        id="teal"
        name="buttongroup"
        onChange={[Function]}
        type="checkbox"
        value="teal"
      />
    </label>
    <label
      className="ButtonGroup-button-0-1-5"
    >
      <span
        className="ButtonGroup-spanText-0-1-6"
      >
        Brown
      </span>
      <input
        checked={false}
        className="ButtonGroup-hiddenInput-0-1-4"
        id="brown"
        name="buttongroup"
        onChange={[Function]}
        type="checkbox"
        value="brown"
      />
    </label>
    <label
      className="ButtonGroup-button-0-1-5"
    >
      <span
        className="ButtonGroup-spanText-0-1-6"
      >
        Black
      </span>
      <input
        checked={false}
        className="ButtonGroup-hiddenInput-0-1-4"
        id="black"
        name="buttongroup"
        onChange={[Function]}
        type="checkbox"
        value="black"
      />
    </label>
    <label
      className="ButtonGroup-button-0-1-5"
    >
      <span
        className="ButtonGroup-spanText-0-1-6"
      >
        Pink
      </span>
      <input
        checked={false}
        className="ButtonGroup-hiddenInput-0-1-4"
        id="pink"
        name="buttongroup"
        onChange={[Function]}
        type="checkbox"
        value="pink"
      />
    </label>
  </div>
</div>
`;
